include/mp/util.h

```
@@ -157,7 +157,7 @@ struct DestructorCatcher
     {
     }
     ~DestructorCatcher() noexcept try {
-    } catch (const kj::Exception& e) {
+    } catch (const kj::Exception& e) { // NOLINT(bugprone-empty-catch)
     }
```

From https://clang.llvm.org/extra/clang-tidy/checks/bugprone/empty-catch.html

> To avoid these issues, developers should always handle exceptions properly

This warning looks legit. I wonder if it can be resolved instead of suppressed.

I am unable to reproduce it, so I am probably doing something wrong. But this should resolve it, suggested from https://github.com/capnproto/capnproto/issues/553#issuecomment-328554603:

```diff
--- i/include/mp/util.h
+++ w/include/mp/util.h
@@ -6,12 +6,13 @@
 #define MP_UTIL_H
 
 #include <capnp/schema.h>
 #include <cstddef>
 #include <functional>
 #include <future>
+#include <kj/debug.h>
 #include <kj/common.h>
 #include <kj/exception.h>
 #include <kj/string-tree.h>
 #include <memory>
 #include <string.h>
 #include <string>
@@ -154,13 +155,14 @@ struct DestructorCatcher
     T value;
     template <typename... Params>
     DestructorCatcher(Params&&... params) : value(kj::fwd<Params>(params)...)
     {
     }
     ~DestructorCatcher() noexcept try {
-    } catch (const kj::Exception& e) { // NOLINT(bugprone-empty-catch)
+    } catch (const kj::Exception& e) {
+        KJ_LOG(ERROR, e);
     }
 };
 
 //! Wrapper around callback function for compatibility with std::async.
 //!
 //! std::async requires callbacks to be copyable and requires noexcept
```

Further, unrelated to this PR, the `DestructorCatcher` does not seem to work in the first place:

```cpp
class A
{
public:
    ~A()
    {
        throw std::runtime_error{"aaa"};
    }
};

template <typename T>
struct DestructorCatcher
{
    T value;
    template <typename... Params>
        DestructorCatcher(Params&&... params) : value(std::forward<Params>(params)...)
    {
    }
    ~DestructorCatcher() noexcept try {
    } catch (const std::runtime_error& e) {
        // Does not work with or without a return.
        //return;
    }
};

int main(int, char**)
{
    try {
        std::shared_ptr<DestructorCatcher<A>> p{std::make_shared<DestructorCatcher<A>>()};
    } catch (const std::runtime_error& e) {
        std::cout << "in main: " << e.what() << "\n";
    }
    return 0;
}
```

```
  * frame #0: 0x0000000825ade18a libc.so.7`__sys_thr_kill at thr_kill.S:4
    frame #1: 0x0000000825a575a4 libc.so.7`__raise(s=6) at raise.c:50:10
    frame #2: 0x0000000825b0ab29 libc.so.7`abort at abort.c:64:8
    frame #3: 0x0000000000203c0e t`__clang_call_terminate + 14
    frame #4: 0x0000000000204e09 t`A::~A(this=0x00000e1f0581a038) at t.cc:184:15
    frame #5: 0x0000000000204db5 t`DestructorCatcher<A>::~DestructorCatcher(this=0x00000e1f0581a038) at t.cc:197:5
    frame #7: 0x0000000000204d59 t`void std::__1::allocator_traits<std::__1::allocator<DestructorCatcher<A>>>::destroy[abi:de180100]<DestructorCatcher<A>, void, void>((null)=0x00000008205e8697, __p=0x00000e1f0581a038) at allocator_traits.h:316:5
    frame #12: 0x0000000000203bf8 t`std::__1::shared_ptr<DestructorCatcher<A>>::~shared_ptr[abi:de180100](this=0x00000008205e8770) at shared_ptr.h:648:17
    frame #13: 0x0000000000203aaa t`main((null)=1, (null)=0x00000008205e8818) at t.cc:206:5
```

---

re: https://github.com/chaincodelabs/libmultiprocess/pull/144#discussion_r1939618932

Good catch. The workaround suggested https://github.com/capnproto/capnproto/issues/553#issuecomment-328554603 is not implemented correctly here. The derived constructor is trying to catch exceptions from a base constructor which will not even run until after it exits. To be implemented correctly it would have to add T as a member instead of deriving from it, and either set T to null or make it a std::optional member to make sure its destruct wont throw after DestructorCatcher exits.

But I think a better fix for this issue is to change [EventLoop::post](https://github.com/chaincodelabs/libmultiprocess/blob/1e06ff07cdeadc78b2c022750f7b5cbc88d56b1d/include/mp/proxy-io.h#L147) method to use [`kj::Function`](https://github.com/capnproto/capnproto/blob/v2/kjdoc/tour.md#functions) instead of `std::function` so `DestructorCatcher` and `AsyncCallable` classes can just be deleted.

---

Now that I can reproduce, I confirm that the above patch which adds `KJ_LOG()` silences the warning and I think that is better than suppressing.

@ryanofsky, I am confused. Do you mean "destructor" instead of "constructor"? Also, there is no inheritance here, so there is no _derived_ constructor. I tried to change `T value;` in the standalone prog above to `std::unique_ptr<T> value;`, `std::optional<T> value;` or to `T* value;` and changed the `~DestructorCatcher()` to:

```cpp
      ~DestructorCatcher() noexcept
      {
          try {
              delete value;
          } catch (const std::runtime_error& e) {
              std::cout << e.what() << "\n";
          }
      }
```
but it keeps crashing.

> But I think a better fix for this issue is to change [EventLoop::post](https://github.com/chaincodelabs/libmultiprocess/blob/1e06ff07cdeadc78b2c022750f7b5cbc88d56b1d/include/mp/proxy-io.h#L147) method to use [kj::Function](https://github.com/capnproto/capnproto/blob/v2/kjdoc/tour.md#functions)

:+1:

---

> ... change [EventLoop::post](https://github.com/chaincodelabs/libmultiprocess/blob/1e06ff07cdeadc78b2c022750f7b5cbc88d56b1d/include/mp/proxy-io.h#L147) method to use [kj::Function](https://github.com/capnproto/capnproto/blob/v2/kjdoc/tour.md#functions) ...

Quick attempt at doing that, I am not sure how to resolve the compilation error in `sync()`:

<details>
<summary>[patch] Remove DestructorCatcher and AsyncCallable</summary>

```diff
diff --git i/include/mp/proxy-io.h w/include/mp/proxy-io.h
index 4430a42..14d5906 100644
--- i/include/mp/proxy-io.h
+++ w/include/mp/proxy-io.h
@@ -8,12 +8,13 @@
 #include <mp/proxy.h>
 #include <mp/util.h>
 
 #include <mp/proxy.capnp.h>
 
 #include <capnp/rpc-twoparty.h>
+#include <kj/function.h>
 
 #include <assert.h>
 #include <functional>
 #include <optional>
 #include <map>
 #include <memory>
@@ -141,20 +142,22 @@ public:
     //! called once from the m_thread_id thread. This will block until
     //! the m_num_clients reference count is 0.
     void loop();
 
     //! Run function on event loop thread. Does not return until function completes.
     //! Must be called while the loop() function is active.
-    void post(const std::function<void()>& fn);
+    void post(kj::Function<void()>& fn);
 
     //! Wrapper around EventLoop::post that takes advantage of the
     //! fact that callable will not go out of scope to avoid requirement that it
     //! be copyable.
     template <typename Callable>
     void sync(Callable&& callable)
     {
+        // XXX
+        // error: no viable conversion from 'reference_wrapper<(lambda at libmultiprocess/include/mp/proxy-io.h:427:43)>' to 'kj::Function<void ()>'
         return post(std::ref(callable));
     }
 
     //! Start asynchronous worker thread if necessary. This is only done if
     //! there are ProxyServerBase::m_impl objects that need to be destroyed
     //! asynchronously, without tying up the event loop thread. This can happen
@@ -192,13 +195,13 @@ public:
 
     //! Handle of an async worker thread. Joined on destruction. Unset if async
     //! method has not been called.
     std::thread m_async_thread;
 
     //! Callback function to run on event loop thread during post() or sync() call.
-    const std::function<void()>* m_post_fn = nullptr;
+    kj::Function<void()>* m_post_fn = nullptr;
 
     //! Callback functions to run on async thread.
     CleanupList m_async_fns;
 
     //! Pipe read handle used to wake up the event loop thread.
     int m_wait_fd = -1;
diff --git i/include/mp/type-context.h w/include/mp/type-context.h
index 7c12afe..40d7cc1 100644
--- i/include/mp/type-context.h
+++ w/include/mp/type-context.h
@@ -61,13 +61,13 @@ auto PassField(Priority<1>, TypeList<>, ServerContext& server_context, const Fn&
 {
     const auto& params = server_context.call_context.getParams();
     Context::Reader context_arg = Accessor::get(params);
     auto future = kj::newPromiseAndFulfiller<typename ServerContext::CallContext>();
     auto& server = server_context.proxy_server;
     int req = server_context.req;
-    auto invoke = MakeAsyncCallable(
+    auto invoke =
         [fulfiller = kj::mv(future.fulfiller),
          call_context = kj::mv(server_context.call_context), &server, req, fn, args...]() mutable {
                 const auto& params = call_context.getParams();
                 Context::Reader context_arg = Accessor::get(params);
                 ServerContext server_context{server, call_context, req};
                 bool disconnected{false};
@@ -140,13 +140,13 @@ auto PassField(Priority<1>, TypeList<>, ServerContext& server_context, const Fn&
                 {
                     server.m_context.connection->m_loop.sync([&]() {
                         auto fulfiller_dispose = kj::mv(fulfiller);
                         fulfiller_dispose->reject(kj::mv(*exception));
                     });
                 }
-            });
+            };
 
     // Lookup Thread object specified by the client. The specified thread should
     // be a local Thread::Server object, but it needs to be looked up
     // asynchronously with getLocalServer().
     auto thread_client = context_arg.getThread();
     return server.m_context.connection->m_threads.getLocalServer(thread_client)
diff --git i/include/mp/util.h w/include/mp/util.h
index 0569c44..4fdfd4e 100644
--- i/include/mp/util.h
+++ w/include/mp/util.h
@@ -6,12 +6,13 @@
 #define MP_UTIL_H
 
 #include <capnp/schema.h>
 #include <cstddef>
 #include <functional>
 #include <future>
+#include <kj/debug.h>
 #include <kj/common.h>
 #include <kj/exception.h>
 #include <kj/string-tree.h>
 #include <memory>
 #include <string.h>
 #include <string>
@@ -143,52 +144,12 @@ template <typename Lock, typename Callback>
 void Unlock(Lock& lock, Callback&& callback)
 {
     UnlockGuard<Lock> unlock(lock); // NOLINT(misc-const-correctness)
     callback();
 }
 
-//! Needed for libc++/macOS compatibility. Lets code work with shared_ptr nothrow declaration
-//! https://github.com/capnproto/capnproto/issues/553#issuecomment-328554603
-template <typename T>
-struct DestructorCatcher
-{
-    T value;
-    template <typename... Params>
-    DestructorCatcher(Params&&... params) : value(kj::fwd<Params>(params)...)
-    {
-    }
-    ~DestructorCatcher() noexcept try {
-    } catch (const kj::Exception& e) { // NOLINT(bugprone-empty-catch)
-    }
-};
-
-//! Wrapper around callback function for compatibility with std::async.
-//!
-//! std::async requires callbacks to be copyable and requires noexcept
-//! destructors, but this doesn't work well with kj types which are generally
-//! move-only and not noexcept.
-template <typename Callable>
-struct AsyncCallable
-{
-    AsyncCallable(Callable&& callable) : m_callable(std::make_shared<DestructorCatcher<Callable>>(std::move(callable)))
-    {
-    }
-    AsyncCallable(const AsyncCallable&) = default;
-    AsyncCallable(AsyncCallable&&) = default;
-    ~AsyncCallable() noexcept = default;
-    ResultOf<Callable> operator()() const { return (m_callable->value)(); }
-    mutable std::shared_ptr<DestructorCatcher<Callable>> m_callable;
-};
-
-//! Construct AsyncCallable object.
-template <typename Callable>
-AsyncCallable<std::remove_reference_t<Callable>> MakeAsyncCallable(Callable&& callable)
-{
-    return std::move(callable);
-}
-
 //! Format current thread name as "{exe_name}-{$pid}/{thread_name}-{$tid}".
 std::string ThreadName(const char* exe_name);
 
 //! Escape binary string for use in log so it doesn't trigger unicode decode
 //! errors in python unit tests.
 std::string LogEscape(const kj::StringTree& string);
diff --git i/src/mp/proxy.cpp w/src/mp/proxy.cpp
index ca094e3..5b2fdf6 100644
--- i/src/mp/proxy.cpp
+++ w/src/mp/proxy.cpp
@@ -19,12 +19,13 @@
 #include <future>
 #include <kj/async-io.h>
 #include <kj/async.h>
 #include <kj/common.h>
 #include <kj/debug.h>
 #include <kj/exception.h>
+#include <kj/function.h>
 #include <kj/memory.h>
 #include <map>
 #include <memory>
 #include <mutex>
 #include <stddef.h>
 #include <stdexcept>
@@ -215,13 +216,13 @@ void EventLoop::loop()
     KJ_SYSCALL(::close(post_fd));
     std::unique_lock<std::mutex> lock(m_mutex); // NOLINT(misc-const-correctness)
     m_wait_fd = -1;
     m_post_fd = -1;
 }
 
-void EventLoop::post(const std::function<void()>& fn)
+void EventLoop::post(kj::Function<void()>& fn)
 {
     if (std::this_thread::get_id() == m_thread_id) {
         fn();
         return;
     }
     std::unique_lock<std::mutex> lock(m_mutex);
```
</details>

---

Thanks for doing ahead with this!

> +        // XXX
> +        // error: no viable conversion from 'reference_wrapper<(lambda at libmultiprocess/include/mp/proxy-io.h:427:43)>' to 'kj::Function<void ()>'
>          return post(std::ref(callable));

This is just what comes to mind but you might be able to replace std::ref with kj::mv. Intent of std::ref is just to avoid a copy.

---

> Now that I can reproduce, I confirm that the above patch which adds `KJ_LOG()` silences the warning and I think that is better than suppressing.

I'd slightly prefer to explicitly suppress the error so we can know to revisit and fix this later than to add KJ_LOG in code we know is broken and will never run.

> @ryanofsky, I am confused. Do you mean "destructor" instead of "constructor"? Also, there is no inheritance here, so there is no _derived_ constructor.

Ooops, I'm not sure what I was looking at when I saw inheritance. And I did mean constructor.

I'm not sure about what was causing crashes you were seeing, but I if we wanted to implement suggested workaround correctly, easiest way would probably be to make member `std::optional<T> value` and call `value.reset()` in the try block.

---

Interesting. In that case could maybe an appealing alternative could be declare T in inside an anonymous union and expliict placement-new to call constructor, and explicit call to its destructor in a try/catch. An approach like that might be able to work if can't get rid of asynccallable class, though I'm still hoping we could get rid of it using kj::Function

---

re: https://github.com/chaincodelabs/libmultiprocess/pull/144#discussion_r1941178169

> [patch] Remove DestructorCatcher and AsyncCallable

This patch is now part of #160